pulsar-metadata/src/main/java/org/apache/pulsar/metadata/impl/AbstractMetadataStore.java

```diff
@@ -317,7 +346,33 @@ public void invalidateAll() {
      */
     protected void execute(Runnable task, CompletableFuture<?> future) {
         try {
-            executor.execute(task);
+            // Wrap the original task, so we can record the thread on which it is running
+            TaskWrapper taskWrapper = new TaskWrapper(task);
+            executorWatchDog.execute(() -> {
```

Comment:
I think that having a wrapper on each task is not the best option here, since we have a single thread where we're making all the blocking calls on the future. 

Also, a 5 delay in a metadata operation does not necessary indicate that there is a deadlock (eg: a particular request could be slow while there is a leader election happening).

Instead, we could have the watchdog thread to execute a dummy task in the `metadata-store` executor. If that task cannot be scheduled for, say, 30sec, then we can be quite confident that there is indeed a deadlock, in which case we should dump the stack traces and kill the broker, since there's no way for it to fix itself up.

Reply:
What you propose also works, except that it is not easy to figure out which code path(task) is the one that is actually blocked. Dumping the stack trace of all threads to logs could result in large volume of logs that one has to wade through to get to the task that is blocked. We could dump the stack trace of all threads to a file, but need alternate mechanisms to extract the file from the server.

One advantage with monitoring each task is that you get the dump of the actual task that is potentially blocked. The 5 seconds is an arbitrary interval, we can change it to 10s or more (whatever we feel is reasonable). Also, note that at this time, we are only logging the stack trace. We could kill the broker if the task exceeds more than say 30s, but I did not do it because I thought that one might want to do that optionally (I can add a knob if we agree that that is what we want to do).

Can you please explain what is the problem with tracking each task? The callback executor is already a single-threaded, which can only execute one task at a time. We can make the monitoring executor multi-threaded but that doesn't help as long as the callback thread is single-threaded.

Reply:
> Can you please explain what is the problem with tracking each task? 

The problem is that we can have 10s or 100s of thousand of such tasks per each second, each with different durations. This adds an overhead in submitting the tasks to the watcher thread. 

With the other approach, we only need to submit a "dummy" task every few seconds to the executor and check that it can complete. It is also completely independent from the timeout of such operations (eg: even if a write operation takes 30seconds, the thread wouldn't normally get blocked for such long time).

So, I think it's much more efficient to track 1 dummy task every few seconds rather than every single task.

> Dumping the stack trace of all threads to logs could result in large volume of logs

When you get the stack trace for the task that is blocked, you're only seeing the stack trace of a healthy code path, while you won't see the other threads that have caused the deadlock. 
Most likely the `metadata-store` thread stack trace will be the most relevant one, though not necessarily.

Reply:
Agreed on we should dump the stack trace of all threads. 
In previous deadlock cases, we got deadlock because metadata-store thread depends on it self. But it's possible that, the depend path is like `metadata-store thread` --> async task in other pulsar-executor --> `metadata-store thread`.  So we need to see the whole picture.

Futher more, I think just dumping the stack trace is not enough, We need the lock info of these threads to help locating the deadlock.  We can use `ThreadDumpUtil#buildThreadDiagnosticString` metioned in #13124

Reply:
w.r.t your comment: "even if a write operation takes 30seconds, the thread wouldn't normally get blocked for such long time".

Just want to clarify my own understanding. Today, the metadata-store callbacks are invoked by the call to execute method:

https://github.com/apache/pulsar/blob/master/pulsar-metadata/src/main/java/org/apache/pulsar/metadata/impl/AbstractMetadataStore.java#L318

And the executor is a single-threaded executor. So, the callbacks can only be executed one at a time. Like you point out there could be 10s or 100s of such tasks.

With my changes, the additional overhead is that the task being submitted to the watchdog thread, which in turn delegates to the metadata-store executor. So, there is a "wrapper task creation overhead" and additional latency latency. But that sholuld not be too bad, compared to the current code? The write operation could take as long, but the watchdog executor is not blocked. It only gets blocked if the callback task gets blocked (which is the current code as well). Is my understanding correct?

And regarding this comment: "When you get the stack trace for the task that is blocked, you're only seeing the stack trace of a healthy code path, while you won't see the other threads that have caused the deadlock."

While I can't comment on all possible code paths, the specific one in the issue 

https://github.com/apache/pulsar/issues/12929

The metadata-store thread stack trace is the most relevant one and that is the only one that gets logged in the code I added.

If you still think that "submitting a dummy task periodically" is the way to go, I will go work on those changes. But please help clarify the above.

Reply:
@Jason918 thank you. Makes sense, somehow I missed thinking about transitive dependencies.

Reply:
@bharanic-dev #13124 has been merged so you could use the added ThreadDumpUtil in this PR.